### ISSUE TITLE
cmd/cue: `get go` can leave stale CUE files behind

### ISSUE THREAD

**Issue by mvdan** [cue-lang/cue#2303]
### What version of CUE are you using (`cue version`)?

<pre>
$ cue version
cue version v0.6.0-0.dev.0.20230320183351-d16b5758c12d

go version devel go1.21-5f1a0320b9 Tue Mar 21 13:10:36 2023 +0000
      -buildmode exe
       -compiler gc
        -ldflags -w -s
  DefaultGODEBUG panicnil=1
     CGO_ENABLED 1
          GOARCH amd64
            GOOS linux
         GOAMD64 v3
</pre>

### Does this issue reproduce with the latest stable release?

Yes.

### What did you do?

I'm using `cue get go` to generate CUE files from Go type definitions across multiple Go packages, like:

```
//go:generate go run cuelang.org/go/cmd/cue get go --local --exclude "^[a-z]" ./...
```

For each Go file with exported Go types, like `foo.go`, I get a file like `foo_go_gen.cue`.

If I move `foo.go` to `bar.go`, `bar_go_gen.cue` appears as expected, but `foo_go_gen.cue` remains in place.

### What did you expect to see?

I expected `cue get go` to not leave its stale generated files behind.

For example, whenever it generates CUE for a Go package, it could start by removing all `*_go_gen.cue` files. Those that are still needed will be created again, and any other will stay removed.

One alternative would be to always generate one CUE file per Go package, like `go_gen.cue`, meaning that we don't have to worry about generated files needing to be removed. Though this would be a slightly more aggressive change.

### What did you see instead?

The above.

cc @rogpeppe

**Comment by myitcv** [comment]
> I expected `cue get go` to not leave its stale generated files behind.

Hmm, I'm not clear that code generators in general should remove files. 

Indeed I can't think of a situation where that does happen. 

Am I missing some obvious examples?

If so, do they remove files by default, or do they do so behind a flag?

**Comment by mvdan** [comment]
In a way, you're right. For example, https://pkg.go.dev/golang.org/x/tools/cmd/stringer produces one generated file per Go type, and doesn't delete anything. However, I have to run it like `stringer -type Foo` to have it generate `foo_string.go`, so I am already in control of which files it is generating, and those filenames are unlikely to change.

Another common example is protobuf generating Go code; it generates one output file per input proto file. Those can be placed in a separate output directory, but one could also have the generated files alongside the source files in the same directory. This makes it closer to `cue get go`, I think, and it also doesn't seem to remove files. Presumably it has the same quirk, though - if I rename `foo.proto` to `bar.proto`, then I can end up with `bar.pb.go` as well as the stale `foo.pb.go`. The difference is that the stale file will be extremely obvious - Go will reject duplicate declarations, for example. `cue get go` often doesn't make that kind of duplication as obvious, as declaring the same definition with the same value twice is OK.

In short - yes, other generators don't delete files, but other generators that I'm familiar with either generate just one file at a time, or often result in errors when stale files are left behind.

**Comment by myitcv** [comment]
I'd agree that it's good practice to remove all generated files prior to running a generator. But in the general case I'm not clear that the generator can do that safely by itself. Especially if one can specify the file into which the generated output is written (because then there ceases to be a pattern of files that it would remove).

**Comment by mvdan** [comment]
I'm going to close this as "unfortunate but not sure we can do better", because I can't disagree with you. I can always search the issue tracker and come back here if I get bitten by this again and have more useful thoughts.
